### PR TITLE
[KHM] Implemented Fire Giant's Fury

### PR DESCRIPTION
**Pull request by weirddan455**
Implemented Fire Giant's Fury for #7248 

The two effect classes are based on code from Light up the Stage.

### CODE REVIEW

File: Mage.Sets/src/mage/cards/f/FireGiantsFury.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        effect.setText("Target Giant you control gets +2/+2");
+        this.getSpellAbility().addEffect(effect);
+
+        effect = new GainAbilityTargetEffect(TrampleAbility.getInstance(), Duration.EndOfTurn);
+        effect.setText("and gains trample until end of turn");
+        this.getSpellAbility().addEffect(effect);
+
+        // Whenever it deals combat damage to a player this turn, exile that many cards from the top of your library. Until the end of your next turn, you may play those cards.
+        Ability ability = new DealsCombatDamageToAPlayerTriggeredAbility(new FireGiantsFuryEffect(), false, true);
+        effect = new GainAbilityTargetEffect(ability, Duration.EndOfTurn);
```
Comment (theelk801) [inline Mage.Sets/src/mage/cards/f/FireGiantsFury.java:48]: This should be implemented as a delayed triggered ability, not a gained ability

**Review by theelk801 — CHANGES_REQUESTED** [review]

File: Mage.Sets/src/mage/cards/f/FireGiantsFury.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    @Override
+    public boolean checkEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.DAMAGED_PLAYER;
+    }
+
+    @Override
+    public boolean checkTrigger(GameEvent event, Game game) {
+        return event.getSourceId().equals(target);
```
Comment (weirddan455) [inline Mage.Sets/src/mage/cards/f/FireGiantsFury.java:113]: I've got it re-implemented as a delayed triggered ability now.  Only problem I found in testing is that it now applies to both combat and non-combat damage (ex. Sorcerer's Wand.)

I tried `return (event.getSourceId().equals(target) && event.getFlag());` here (saw a comment in GameEvent.Java that said flag should be true for combat damage) but doing this just stopped the ability from triggering entirely.

Comment (theelk801) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java:113]: You have to cast it to `DamagedEvent` and use `isCombatDamage()`

Comment (theelk801) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java:113]: Also you don't need the watcher, you can pass the damage amount to the effect in the `checkTrigger` method

File: Mage.Sets/src/mage/cards/f/FireGiantsFury.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    public FireGiantsFury copy() {
+        return new FireGiantsFury(this);
+    }
+}
+
+class FireGiantsFuryEffect extends OneShotEffect {
+
+    public FireGiantsFuryEffect() {
+        super(Outcome.PlayForFree);
+        this.staticText = "Whenever it deals combat damage to a player this turn, exile that many cards from the top of your library. Until the end of your next turn, you may play those cards";
```
Comment (theelk801) [inline Mage.Sets/src/mage/cards/f/FireGiantsFury.java:71]: You should add `<br>` to the beginning of this to add a line break.

**Review by theelk801 — CHANGES_REQUESTED** [review]

File: Mage.Sets/src/mage/cards/f/FireGiantsFury.java (hunk 3 above)
Comment (JayDi85) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java:71]: It's a one line ability in printed card. So no needs in line break.

Comment (theelk801) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java:71]: Oh ok then scratch that

Comment (weirddan455) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java:71]: Thanks.  I got it to check for combat damage correctly now.  However, I can't seem to pass the damage in `checkTrigger` like you said.  I made a damage variable:

```
class FireGiantsFuryDelayedEffect extends OneShotEffect {

    private int damage = 0
```

Created a setDamage method:

```
public void setDamage(int damage) {
    this.damage = damage;
}
```

Then call it from `checkTrigger`:

```
public boolean checkTrigger(GameEvent event, Game game) {
    if (event.getSourceId().equals(target) && ((DamagedEvent) event).isCombatDamage()) {
        FireGiantsFuryDelayedEffect effect = (FireGiantsFuryDelayedEffect) this.getEffects().get(0);
        effect.setDamage(event.getAmount());
        return true;
```

I added some debugging code and `setDamage` is being called and is being passed the correct number but for some reason when `apply` runs and checks `damage`, the value is 0.

I feel like I'm missing something simple here but having trouble figuring out what.

Comment (theelk801) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java:71]: take a look at some of the other combat damage triggered abilities

Comment (weirddan455) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java:71]: Do you know of any example cards that use a delayed trigger on combat damage?  I looked at the code for Garruk's Harbinger and it just does

`Integer damage = (Integer) getValue("damage");`

inside `apply` but this doesn't work for a delayed trigger.

I'm coming up empty searching ScryFall for cards with similar wording.  The card seems somewhat unique in that way.

The watcher I made does work though.  Is there a reason why that's not the correct way to handle this?

Comment (theelk801) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java:71]: The problem with the watcher is it doesn't work if some of the damage is redirected to a different target. Take a look at the `checkTrigger` method for `DealsCombatDamageToAPlayerTriggeredAbility`.

Comment (JayDi85) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java:71]: [Hunter’s Insight](https://scryfall.com/card/cmr/232/hunters-insight)

Choose target creature you control. Whenever that creature deals combat damage to a player or planeswalker this turn, draw that many cards.

Comment (theelk801) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java:71]: yeah, that way works too

Comment (weirddan455) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java:71]: Oh, thanks that was helpful.  I think I've got it working now.  Let me just clean up the code and I'll push another commit with the changes.

File: Mage.Sets/src/mage/cards/f/FireGiantsFury.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    public boolean apply(Game game, Ability source) {
+        DelayedTriggeredAbility delayedAbility = new FireGiantsFuryDelayedTriggeredAbility(source.getFirstTarget());
+        game.addDelayedTriggeredAbility(delayedAbility, source);
+        return true;
+    }
+}
+
+class FireGiantsFuryDelayedTriggeredAbility extends DelayedTriggeredAbility {
+
+    private final UUID target;
```
Comment (theelk801) [inline Mage.Sets/src/mage/cards/f/FireGiantsFury.java:91]: this should be a `MageObjectReference` so you can track whether it's left the battlefield and returned, take a look at Hunter's Insight for reference.

**Review by theelk801 — CHANGES_REQUESTED** [review]

File: Mage.Sets/src/mage/cards/f/FireGiantsFury.java (hunk 4 above)
Comment (weirddan455) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java]: Yea, I was looking at Searing Blood and Reincarnation when I wrote this that use UUID like this in their delayed triggered abilities but this is probably better.  I changed it.

**Review by theelk801 — APPROVED** [review]

File: Mage.Sets/src/mage/cards/f/FireGiantsFury.java (hunk 4 above)
Comment (theelk801) [reply Mage.Sets/src/mage/cards/f/FireGiantsFury.java]: There are definitely older cards that need to be refactored, it's an unending battle.

**Pull request merged**